tony-core/src/main/java/com/linkedin/tony/TonyClient.java

```diff
@@ -756,18 +811,31 @@ private boolean monitorApplication()
       ApplicationReport report = yarnClient.getApplicationReport(appId);
 
       YarnApplicationState state = report.getYarnApplicationState();
+
       FinalApplicationStatus dsStatus = report.getFinalApplicationStatus();
       initRpcClient(report);
 
-      // Query AM for taskUrls if taskUrls is empty.
-      if (amRpcServerInitialized && taskUrls.isEmpty()) {
-        taskUrls = amRpcClient.getTaskUrls();
-        if (!taskUrls.isEmpty()) {
-          if (callbackHandler != null) {
-            callbackHandler.onTaskUrlsReceived(ImmutableSet.copyOf(taskUrls));
+      if (amRpcClient != null) {
+        Set<TaskInfo> receivedInfos = amRpcClient.getTaskInfos();
+        Set<TaskInfo> taskInfoDiff = receivedInfos.stream()
+                .filter(taskInfo -> !taskInfos.contains(taskInfo))
+                .collect(Collectors.toSet());
+        // If task status is changed, invoke callback for all listeners.
+        if (!taskInfoDiff.isEmpty()) {
+          LOG.info(receivedInfos);
+          for (TaskUpdateListener listener : listeners) {
+            listener.onTaskInfosReceived(receivedInfos);
+          }
+        }
+        taskInfos = receivedInfos;
```

Comment:
Why do you need both `taskInfos` and `receivedInfos`?

Reply:
Received info is what you received from last call to AM, taskInfos is the previous call to AM. Here is a comparison between these two and we only invoke the listener if anything has changed

Reply:
Ah, I see. I think this should be moved up one line to inside the `if (!taskInfoDiff.isEmpty())` block? If there is no diff, we don't need to update `taskInfos` I think. (I guess from a correctness standpoint it doesn't matter either way, but I just think logically it makes more sense to only update it if something actually changed.)

Reply:
Makes sense, moved